### PR TITLE
Viewport

### PR DESCRIPTION
**Pull request by biharck**
Adding the ability to use hotkeys with:
```
Default Tool
Zoom
Pan
Angle measurement
Scroll stack
Length measurement
Flip Horizontally
Flip Vertically
Rotate Right
Rotate Left
Invert
Zoom In
Zoom Out
Zoom to Fit
Reset
Clear Tools
Scroll Down
Scroll Up
Scroll to the Last Image
Scroll to First Image
```

### CODE REVIEW

**Comment by dannyrb** [comment]
@biharck, can you remove the `package-lock.json` and add it to your `.gitignore`?

The main OHIF projects use yarn and commit `yarn.lock`

File: src/classes/HotkeysUtil.js [hunk 1]
```diff
@@ -10,18 +10,18 @@ export default class HotkeysUtil {
       zoom: 'Zoom',
       angle: 'Angle',
       dragProbe: 'Pixel Probe',
-      ellipticalRoi: 'Elliptical ROI',
-      rectangleRoi: 'Rectangle ROI',
+      ellipticalRoi: 'EllipticalRoi',
```
Comment (dannyrb) [inline src/classes/HotkeysUtil.js:15]: Is the second value here the display name? If yes, can we match the display name in the meteor version?

image

File: src/classes/HotkeysUtil.js [hunk 2]
```diff
@@ -31,52 +31,164 @@ export default class HotkeysUtil {
...
+
+  _rotate(currentPosition, rotationDegree) {
+    return currentPosition === undefined
+      ? 0 + rotationDegree
+      : currentPosition + rotationDegree
+  }
+
+  _dispatchCommand(options) {
+    const { setActiveViewportSpecificData } = actions
+    window.store.dispatch(
```
Comment (dannyrb) [inline src/classes/HotkeysUtil.js]: Is this the best/only way to access the store?

File: src/redux/constants/ActionTypes.js [hunk 3]
```diff
@@ -5,3 +5,4 @@ export const SET_VIEWPORT = 'VIEWPORT::SET'
 export const SET_VIEWPORT_ACTIVE = 'VIEWPORT::SET_ACTIVE'
 export const SET_VIEWPORT_LAYOUT = 'VIEWPORT::SET_LAYOUT'
 export const CLEAR_VIEWPORT = 'VIEWPORT::CLEAR'
+export const SET_VIEWPORT_SPECIFIC_DATA = 'VIEWPORT::SET_VIEWPORT_SPECIFIC_DATA'
```
Comment (dannyrb) [inline src/redux/constants/ActionTypes.js]: To match the others, can we make the string value: `VIEWPORT::SET_SPECIFIC_DATA`

Where `VIEWPORT` is the scope/module, and what comes after the `::` is the action

File: src/redux/reducers/viewports.js [hunk 4]
```diff
@@ -48,6 +50,28 @@ const viewports = (state = defaultState, action) => {
...
+        cloneDeep(state.viewportSpecificData[action.viewportIndex]) || {}
+      viewportSpecificData = cloneDeep(state.viewportSpecificData)
+      viewportSpecificData[action.viewportIndex] = merge(
+        {},
+        currentData,
+        action.data
+      )
+
+      return Object.assign({}, state, { viewportSpecificData })
+    case 'SET_ACTIVE_VIEWPORT_SPECIFIC_DATA':
```
Comment (dannyrb) [inline src/redux/reducers/viewports.js]: - Can we add this to ActionTypes?
- Can we chain actions? IE. this calls `SET_VIEWPORT_SPECIFIC_DATA`, but passes the `activeViewportIndex`?

**Comment by dannyrb** [comment]
You added a new Action, which broke a unit test. Can you update the unit test to include the new Action? You can see the error by clicking "Details" next to the integration check.

image


It should take you here:

https://circleci.com/gh/OHIF/ohif-core/70?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

File: src/classes/HotkeysUtil.js (hunk 2 above)
Comment (biharck) [reply src/classes/HotkeysUtil.js]: Actually, I am not 100% sure about it either. We can catch up in order to design the best approach.

File: src/classes/HotkeysUtil.js (hunk 1 above)
Comment (biharck) [reply src/classes/HotkeysUtil.js:15]: Not it is not. The display name comes from `reducers/preferences.js` https://github.com/OHIF/ohif-core/blob/master/src/redux/reducers/preferences.js#L5

File: src/redux/constants/ActionTypes.js (hunk 3 above)
Comment (biharck) [reply src/redux/constants/ActionTypes.js]: done

File: src/redux/reducers/viewports.js (hunk 4 above)
Comment (biharck) [reply src/redux/reducers/viewports.js]: done

File: src/classes/HotkeysUtil.js (hunk 2 above)
Comment (evren217) [reply src/classes/HotkeysUtil.js]: @biharck we should never use `window.store` in this package.

We should get a callback and trigger that callback, so let the caller decide what to do with it. See how [Hanging Protocol Engine](https://github.com/OHIF/ohif-core/blob/63690998e069c5f2027841a08485afd643aebe00/src/hanging-protocols/ProtocolEngine.js#L693) do that.

**Comment by dannyrb** [comment]
@biharck if you resolve @evren217's issue and the conflict, then I think we can merge this and your other PR. I do believe we will want to change these actions in the (very) near future to use the exposed cornerstone DOM element instead of reacting to state changes though.

**Pull request merged**